deck: Display a heat map of the last 12 hours of job successes/failures

smarterclayton opened the pull request:
When triaging job failures it is common to want to identify correlations in time, which the list of builds not well suited to because it doesn't make quick assessment of the gaps between builds visible.  There are two primary user personas who might want to do this - the user assessing infrastructure problems (`prow admin`) and the user assessing job related problems (`job admin` and `job viewer`).  The job admin wants to know if their changes broke the job, and they almost certainly know the time that their job change went in.  The job viewer wants to know if the job in general is flaky, so they only need a quick glimpse at the total list of jobs.  Deck is replacing gubernator for this sort of listing, so it makes sense for us to ensure that a quick visualization of build status over time exists (vs just build status in aggregate like the job bar provides today).

Use a heat map to display the last 12 hours of the selected runs, newest to oldest, with the
y dimension being the duration of the build, the color of the square being partially filled with green and red depending on the percentage of failures in that square, and the opacity of the square being the percent of jobs in the entire column.  This heap map shows up right below the job bar and uses ~100 vertical pixels to capture a summary of the state of the filtered jobs.

![image](https://www.dropbox.com/s/80n5fahrpvflquj/Screenshot%202019-02-26%2023.57.46.png?raw=1)

The end goal is to make time-correlated failures across multiple jobs more obvious. For instance, looking at a PR job shows a vertical line as all the jobs kick off, while looking at a single job across many PRs can show the exact time when test started regressing, and the user can then scan into
deeper detail.  In the example above you can see outlier builds along the top, indicating there may be a specific failure mode that takes much longer than the others.  You can also assess the percentage of success at any time visually.

Displays the left and right times, then hovering over a cell lets you click to jump straight to that build with a subtle hover effect to suggest interactivity.

smarterclayton commented:
@stevekuznetsov came out of the discussions about trying to identify when a job started to flake - you *can* do this in prometheus and grafana, but it's not as effective for a quick scan, which deck is an excellent place to do that from.

stevekuznetsov commented:
Adding @cjwagner and @fejta in addition to @BenTheElder and @Katharine

fejta commented:
pretty nifty

smarterclayton commented:
I added interactivity - if you hover over a colored square it will scale (so it looks like it's selectable) and if you click it it will move to the first sample in the list (so if you click an outlier you'll scroll to that outlier).  Doesn't work great for the end of the list, BUT since the heat map is 12h it's common to have more things than show up in the heat map.  Could use some more thinking about.

smarterclayton commented:
Tweaked the opacity to be proportional to number of samples per square, fixed a number of bugs, better title text.  This is in a "working state" - open to any feedback folks have.  I played with a bunch of variants and there are some interesting things that could be explored, but this at least stands on its own and doesn't distract too much.

![images](https://www.dropbox.com/s/12jloeq0qgxix1r/Screenshot%202019-02-25%2022.05.53.png?raw=1)

BenTheElder commented:
[this looks nifty -- would like @Katharine to take a look 🙃]

Katharine commented:
I like the concept!

I haven't had a change to look at the code yet, but I've run it locally and I have a few general comments:

* Clicking squares doesn't seem to do anything? Your description suggests it should.
* Showing the time 12 hours ago to indicate the past works poorly [in the afternoon] because the time twelve hours ago is represented the same as the time now [or, at least, users expect it to be]. Just putting "12 hours ago" might be clearer.

I think this should be collapsible, and default to collapsed. Aside from it doubling the amount of header, I don't think it's very useful when just showing all jobs.

By default, on prow.k8s.io, you get this:

<img width="1407" alt="image" src="https://user-images.githubusercontent.com/110792/53394950-8782a400-3955-11e9-989c-5c570d4dac41.png">

This suggests that most tests are failing most of the time, which is actually false! I wonder if we could have a less misleading rendering of passed vs failed tests without defeating the point (maybe a fade between red/green? I haven't tried experimenting with this).

Other miscellaneous thoughts:

* It's worth thinking about the overlap of functionality we have between the job history view, TestGrid, and this
* I would be tempted to suggest that this should maybe be on the job history page, but I am assuming you want this to work across multiple jobs, in which case that would not make sense
* I'd maybe throw in a mention about red/green colourblindness ([it looks like this](https://user-images.githubusercontent.com/110792/53395829-b1d56100-3957-11e9-830e-e127a8c6b60e.png)), but that's a broader problem we have everywhere and is way out of scope here.

smarterclayton commented:
So the question is who goes to deck?  I go to deck to debug why things are failing.  If things are working, I don't go to deck.  Why would I go to deck if my things are passing?

> Clicking squares doesn't seem to do anything? Your description suggests it should.

Because you're in the overview you're capped at 500 builds in the list so clicking can't do anything.  Could make clickability based on whether the build is in the list.

> Showing the time 12 hours ago to indicate the past works poorly [in the afternoon] because the time twelve hours ago is represented the same as the time now [or, at least, users expect it to be]. Just putting "12 hours ago" might be clearer.

Agree.

> I think this should be collapsible, and default to collapsed. Aside from it doubling the amount of header, I don't think it's very useful when just showing all jobs most of the time (perhaps more useful when everything is 🔥🔥🔥).

I don't think it should be collapsed unless we can identify which percentage of users * priority they spend in this list to identify failures or outliers.  Who looks at deck?  As far as I can tell it's only people who fit into the "triage of prow" role in an organization, with a smattering of people who want to filter their jobs.

I would argue the visual / action prioritization of elements in deck should be:

1. job bar - I can see a simplistic overview of total failures and immediately look at failures
2. job histogram - I can see a slightly more detailed overview
3. filter section - I can scope down to what I care about
4. the list - if I've already filtered
....
10000. the list unfiltered in a large project

I don't think default list of items in k8s can be used to reason about to anyone in k8s since it contains too much info to be useful - in smaller projects the entire list might be more useful, but you probably have higher green percentages.  In general I think the goal of deck is to help you:

1. see when something started breaking instead of going to gubernator lists (which histogram is better than list)
2. navigate to a particular job by filtering (in which case the histogram isn't so big that it impacts you)

> This suggests that most tests are failing most of the time, which is actually false!

Given historical precedent in Kube flakes are 90+% (I would say 99.9% based on personal history) of all failed jobs.  I would think deck is most useful to triage and identification of failures.  So I would say the point of the histogram is to help you get back to green or to filter the list.

smarterclayton commented:
> It's worth thinking about the overlap of functionality we have between the job history view, TestGrid, and this

Neither job history view or test grid can currently show you trends across multiple jobs (which is a huge gap for people who need to run the same tests across most repos).  Deck is a better job history view than either.  I can't make changes to test grid until it's open sourced, but once it is I think the sort of action (when did correlated jobs across repos start failing) that we want is possible, just much harder.  I think test grid should be reserved for test flake visualization rather than test job visualization.  Test grid is worse at quick pivot navigation, and is also fairly slow.  

> I would be tempted to suggest that this should maybe be on the job history page, but I am assuming you want this to work across multiple jobs, in which case that would not make sense

Correct.  This is about quick summarization across repos, jobs, statuses.

> I'd maybe throw in a mention about red/green colourblindness ([it looks like this](https://user-images.githubusercontent.com/110792/53395829-b1d56100-3957-11e9-830e-e127a8c6b60e.png)), but that's a broader problem we have everywhere and is way out of scope here.

Yeah, and that's fixable for a user by filtering by failures and then filtering by success.

smarterclayton commented:
Maybe I could summarize the mindset as:

1. Successful jobs are irrelevant - you don't need to prioritize looking at them in the general case
2. Most people replacing gubernator in their workload are going to filter by job instantly - histogram will be useful for them
3. Anyone triaging failures wants the histogram more than the list.

fejta commented:
/assign @Katharine @michelle192837 
LGTM, have minimal JS skills

michelle192837 commented:
(I'll take a look, but just to chime in this is definitely something I'd like to see in TestGrid (at least in the summary) and also this seems good until TestGrid's actually open-sourced 👍 Unless someone else on the Google side wants to take that on.)

The red-green colorblindness does seem like a potential issue; TestGrid relies on a large difference in saturation between statuses, which opacity negates here. If there's a quick fix for that that'd be awesome.

smarterclayton commented:
Is there anything in more recent CSS that can help us?  We could potentially do alternating grid for success and solid for failed.

Katharine commented:
> Because you're in the overview you're capped at 500 builds in the list so clicking can't do anything. Could make clickability based on whether the build is in the list.

I think this is a poor user experience - if we're going to show data an interactive visualisation of data, we should be consistent about when it is interactive. There is no technical constraint on being able to reliably do something, but we would need to devise a different way of displaying information. Perhaps clicking on a cell should act as an additional 'filter' such that we show only the results from the currently-selected cell?

> Who looks at deck? As far as I can tell it's only people who fit into the "triage of prow" role in an organization, with a smattering of people who want to filter their jobs.

This might be true if this were on a dedicated page, but this is in practice the homepage for prow, and is where you get back to if you click the logo at the top of a deck page. In the immediate future (like, today), deck will completely replace gubernator for k8s-project job viewing, so deck in general will see many users.

That said, _regardless_ of the type of people who end up on this page, I don't think it is good for the most visible thing on the page to be a giant pile of red that massively exaggerates the failure rate. Would being clearer actually make this less useful? I'm not convinced it would, but I could believe otherwise

> Yeah, and that's fixable for a user by filtering by failures and then filtering by success.

I don't think that's a reasonable solution, but confess I don't necessarily have a good answer off the bad.

Since this PR is full of LGTMs and "nifty"s, I'm going to be explicit: I like the concept a lot, but I am not in favour of this landing with the current UX: it's visually large, misleading, has surprising behaviour, and assumes a very specific usage pattern that I am not convinced is accurate (it's not mine!).

stevekuznetsov commented:
Can we add a `/triage` page, dupe the part of the the current landing page behavior that is useful, and target the admin use case there?

stevekuznetsov commented:
Deck and Spyglass in general are full of UX nightmares and I think it's generally true that there are distinct types of users and we should try to cater to making pages that are useful to them instead of one that is semi-useful to everyone

smarterclayton commented:
> I think this is a poor user experience - if we're going to show data an interactive visualisation of data, we should be consistent about when it is interactive. There is no technical constraint on being able to reliably do something, but we would need to devise a different way of displaying information. Perhaps clicking on a cell should act as an additional 'filter' such that we show only the results from the currently-selected cell?

I meant - this is a bug.

> This might be true if this were on a dedicated page, but this is in practice the homepage for prow, and is where you get back to if you click the logo at the top of a deck page. In the immediate future (like, today), deck will completely replace gubernator for k8s-project job viewing, so deck in general will see many users.

Right, but my argument is that the majority of those users are the same type of people who are going to filter to a job immediately (at which point histogram is useful).  If this is solely about "I feel bad that the default deck list is red is bad" we can definitely change that.

I have trouble imagining us encouraging the UX point of view that the unfiltered list of prow jobs on the homepage is useful.  In that line of argument, we should hide the list and show the histogram.

> That said, _regardless_ of the type of people who end up on this page, I don't think it is good for the most visible thing on the page to be a giant pile of red that massively exaggerates the failure rate.

I think massively exaggerates is the key point here.  Looking at deck right now:

![image](https://www.dropbox.com/s/6a9e2kjbms9k360/Screenshot%202019-02-26%2014.10.43.png?raw=1)

30% of builds are failing.  I suspect the real failure rate (non-flake related) is more like 5%, at best.  Are we saying that we'd be ok if 30% of the graph was red?  Just not 80%?  It's possible that we can bias heavily when numbers increase.  Certainly when you've filtered the job, I want to see failures.

To get at the "home page for prow" - the pending jobs at the top of the list have 0% informational value for any user except possibly to let them infer a list exists.  It would be better to take 10% of that vertical space and spend it on the histogram than to display 3 rows of pending jobs.

smarterclayton commented:
> ... in favour of this landing with the current UX: it's visually large, misleading, has surprising behaviour, and assumes a very specific usage pattern that I am not convinced is accurate (it's not mine!).

Can you describe what your usage pattern is?  The three that I have in my mental model for prow are:

1. People who today follow a link in from GitHub to a specific PR Gubernator page and in the future use spyglass.  These people only care about the specific PR and are trying to identify why the job failed.  Information about other jobs is irrelevant.  Primary UX focus: reduce time to identify and fix failure.
2. People who want to see the list of builds under a job so they can see the general trend rate of failures.  These people are either debugging when a job got flaky or identifying whether *they* caused the flake.  Primary UX focus: convey trend across enough time (4-24h) and allow easy access to individual results.
3. People who are looking at aggregate status for a repo, a job, or are triaging prow related failures.  Primary UX focus: convey an effective summary of the failure rate over time (the current prow job summary is insufficient because I could be 99% passing and 1% failing and all that matters is whether the 1% is now or not).

For 2-3 the histogram is better than the list of builds - it more succinctly conveys the information about the job, but still allows navigation.

For 1 the histogram is irrelevant, but I assume 99% of users are coming in only from PR links and so they shouldn't even see the prow homepage.

Katharine commented:
> Can we add a `/triage` page, dupe the part of the the current landing page behavior that is useful, and target the admin use case there?

That would be a compelling reason to not bother with the thing being collapsible, but I'm not sure it'd be a compelling argument against anything else.

I _do_ agree that perhaps the deck front page should be something else, though I could not tell you exactly what that might be.

> Can you describe what your usage pattern is?

My primary usage patterns are:

- Is anything working?
  - Here the current histogram is probably misleading because almost everything is some sort of red (and so suggests "no"), but would be useful here with changes. The list also works basically fine here.

- What is up with that job I just kicked off? Does it exist yet?
  - List > histogram here

--------

Again, to be clear, I am in favour of this histogram existing. I think it could use some changes to improve the UX (the smattering of green with the pile of red feels misleading, the behaviour on clicking could be improved), and I am still not necessarily sold that it needs to be consistently visible (but also not super opposed).

Katharine commented:
> Deck and Spyglass in general are full of UX nightmares

I have heard no feedback about Spyglass to this effect, and would be interested in specific complaints (in another venue that is not this PR).

smarterclayton commented:
> Is anything working? Here the current histogram is probably misleading because almost everything is some sort of red (and so suggests "no"), but would be useful here with changes. The list also works basically fine here.

Part of my motivation was doing this was because the list wasn't working well for dealing with partial failures (not just prow is down, but "all bazel builds are failing but all e2e jobs are working" sorts of scenarios).  I don't disagree that list contains the info, but it requires a huge amount of space to convey it - it's information sparse rather than information dense.

It sounds like the use case that you want to capture "is the whole build system impacted", while the primary use case I was targeting was "how do we identify systematic flakes and failures for chunks of jobs".

So what are our options?  The current algorithm is to make the square red if any failure occurred, because reds should be (but aren't) rare.  In aggregate on kube (20k builds over 3/4 days) or openshift (6k over 2 days) that means everything is red because our failure rate is dramatically higher than the real failure rate.  Usability wise I assume we want to make the following actions easy:

1. visually estimate failure percentage from overall histogram within 10-15% (your comment)
2. estimate time based failure regions (things started failing at X)
3. bias towards identifying failures vs identifying successes, but don't drown out successes

I'll take a look at a couple of variants of presentation that preserve 1 without losing 3.

> What is up with that job I just kicked off? Does it exist yet? List > histogram here 

Is this as someone helping add new jobs to Prow?  I.e. general prow maintainer?  I can't say that I've ever gone to deck to find the job when a PR is there as a contributor, but as a maintainer of jobs I would agree that I sometimes do that.  Is that something other prow maintainers would expect to do *without* filtering?

The reason I ask is that the odds of your job being in the top on a busy system (3-10 builds a minute) are already low.  So you are already scrolling slightly - scrolling 3 more lines (rough equivalent height) seems less of an issue.

smarterclayton commented:
On the first page problem, this variant might be a reasonable compromise:

![image](https://www.dropbox.com/s/80n5fahrpvflquj/Screenshot%202019-02-26%2023.57.46.png?raw=1)

I'm using a background gradient with 10 discrete steps (100% success, 90% success, etc down to 0% success) to do sub-block proportionality.  That will give us within 10% the exact amount of green and red.  It doesn't fix the red/green color issue but it does address proportionality.  It has a higher CPU cost (gradients are expensive) but works on major browsers reasonably well.

stevekuznetsov commented:
@Katharine this is beside the point a little bit but I did want to mention that:

 > Is anything working?

The current overview gives you little to no useful information about this. How do you determine that? The failure rate is so high organically it is not unreasonable to see a page of red builds as it is during normal operation. The quick stacked bar chart doesn't help since it has a lot of hysteresis. The overview gives a very, very bad signal for this question and we should not design the overview with this in mind. We should use metrics, logging and alerting to determine the health of the system.
 
> What is up with that job I just kicked off? Does it exist yet?

If you manage to hit the page at a time where your build is at the top, ok, but in every other case it will have been taken out of the top by other activity. Why not filter? Also, in what case are you looking at this but _not_ GitHub? Are you considering a case where you have a `crier` outage? Why not use

```
kubectl get prowjobs --selector prow.k8s.io/job=foo
```

I think my broader point is -- the use cases _you_ seem to prioritize for this page are weakly supported and somewhat unique. I'm not aware of the general body of contributors using the page in that way and administrators have much more powerful and useful tools at their disposal to answer the questions you have. I'm not sure we should be focusing on supporting those specific use cases in this page.

stevekuznetsov commented:
Not to be all project-management-y but I think it would help to flesh out some formal user profiles for Deck so that we can be all on the same page when talking about all of this. If the conversation continues in this thread as to who will use this and how, I propose we talk about:

 - individual contributors writing pull requests
 - super-users or repository/org admins looking at a subset of job health
 - Prow administrators

Katharine commented:
> On the first page problem, this variant might be a reasonable compromise:

I like this version! I'll concede on making it collapsible.

I'd be happy with this + consistent behaviour when clicking on cells. I would also request the animation when hovering be removed or tweaked - it feels more like lag than an intentional transition to me.

spiffxp commented:
Hi so there are large walls of text here, and I have concerns we're blowing scope on the original PR here.

Depending on how you look at it, this can either look like a large change that's been implemented without prior discussion, or a proposal with a working implementation.  I wonder if we could have saved ourselves some back and forth if discussed some of this in-person at the SIG Testing meeting yesterday, which had room on the agenda.

That said, I would like to understand how we can take what looks like a neat idea and iterate forward.  The most recent compromise looks good to me.  If it's still at a place where it doesn't work for all users of prow, maybe it's something we gate behind a feature flag.

We happen to have another tool for triaging kubernetes failures that is not explicitly part of prow, https://go.k8s.io/triage.  It relies on a data pipeline of gcs -> kettle -> bigquery + periodic prowjob to do its thing, but is otherwise a webpage that lives in GCS.

spiffxp commented:
@stevekuznetsov can you take your concerns to followup issues so we can address them out of the scope of this PR?

smarterclayton commented:
> I wonder if we could have saved ourselves some back and forth if discussed some of this in-person at the SIG Testing meeting yesterday, which had room on the agenda.

I’m sure it could - I think Katherine’s feedback actually helped refine down to some fundamental sig testing questions, which as Steve notes we’re better prepared to answer now.

Triage is unfortunately heavy for an external team to run today.  I’d like us to get there, but I do think we should be careful to carefully layer prow into pieces that stand on their own.

As I resolve Katherine’s feedback I’m happy to distill the discussion here into either the PR description or a proposal.

smarterclayton commented:
Ok, updated with the following:

* variant for histogram display described above that shows relative accurate proportional weight for boxes (so 90% success looks more like 90% success rather than 90% failure)
* replaced hover expand behavior with hover now setting opacity to 1 which is more subtle but still implies clickable
* only builds that are in the filtered prowjob list are selectable in the overview (i.e. when we have over 500 only the first 500 items are clickable).  The rest have no hover behavior and appear inert.

PTAL

Katharine commented on prow/cmd/deck/static/prow/prow.ts:
```diff
@@ -674,6 +704,127 @@ function stateToAdj(state: JobState): string {
+    // initialize the table if the row count changes
+    if (el.childNodes.length !== rows) {
+        el.innerHTML = "";
+        for (let i: number = 0; i < rows; i++) {
```
`number` is inferred and can be omitted.

```suggestion
        for (let i = 0; i < rows; i++) {
```

Katharine commented on prow/cmd/deck/static/prow/prow.ts:
```diff
@@ -674,6 +704,127 @@ function stateToAdj(state: JobState): string {
+        el.innerHTML = "";
+        for (let i: number = 0; i < rows; i++) {
+            const tr = document.createElement('tr');
+            for (let j: number = 0; j < cols; j++) {
```
`number` is inferred and can be omitted.

```suggestion
            for (let j = 0; j < cols; j++) {
```

Katharine commented on prow/cmd/deck/static/prow/prow.ts:
```diff
@@ -674,6 +704,127 @@ function stateToAdj(state: JobState): string {
+                td.className = `success-${Math.floor(10 - (failures / samples.length) * 10)}`;
+            }
+        });
+        for (let rowIndex: number = lastRowIndex; rowIndex < rows; rowIndex++) {
```
`number` is inferred here and can be omitted.

```suggestion
        for (let rowIndex = lastRowIndex; rowIndex < rows; rowIndex++) {
```

Katharine commented on prow/cmd/deck/static/prow/prow.ts:
```diff
@@ -697,3 +848,121 @@ function createIcon(iconString: string, tooltip: string = ""): HTMLAnchorElement
+    public end: number;
+    public max: number;
+
+    constructor(data: JobSample[][], start: number, end: number, max: number) {
```
As above, you could also write this constructor like so to repeat yourself less:

```ts
export class JobBuckets {
    constructor(public data: JobSample[][],
                public start: number,
                public end: number,
                public max: number) { }

    // ...
}
```

smarterclayton commented:
Added tests but may need some assistance debugging why bazel chokes with `no specs found` (probably related to directory structure compared to gopherage)?

smarterclayton commented:
Nm, PEBKAC, needed to use `_test.ts` instead of `_tests.ts`.  PTAL
